### PR TITLE
Add unit test for parse.hpp

### PR DESCRIPTION
**Pull request by ayushgupta138**
Unit test for extensions/test/algorithms/parse.hpp for the issue #788

### CODE REVIEW

**Comment by ayushgupta138** [comment]
@mloskot  @vissarion  @awulkiew  could you please review the pr and merge the changes.

**Comment by mloskot** [comment]
@ayushgupta138 I have just merged PR #802 with initial (first) submission of the tests by @jkerkela 

Next, I'd like to combine your tests, which also cover the parser from different angle, into those tests.
Would you be interested to do it?

**Comment by ayushgupta138** [comment]
Yes , I am interested in it. Let me know what you need from my side.

**Comment by mloskot** [comment]
1. Update your `unit_test_parse` branch with the latest changes from `develop`.
2. Resolve any conflicts - you will get some, obviously, as `extensions/test/algorithms/parse.cpp` is a new file added by both, #802 and #806
3. Copy/Add your test cases to the `extensions/test/algorithms/parse.cpp`
4. `git push --force` to update this PR

**Comment by ayushgupta138** [comment]
On it.

**Comment by ayushgupta138** [comment]
As mentioned by you, I made commits on this pr.

File: extensions/test/algorithms/Jamfile [hunk 1]
```diff
@@ -12,6 +12,7 @@ test-suite boost-geometry-extensions-algorithms
     :
 #    [ run dissolve.cpp ]
     [ run distance_info.cpp ]
+    [ run parse.cpp]
```
Comment (mloskot) [inline extensions/test/algorithms/Jamfile:15]: This is no longer necessary since parse.cpp has already been added, in #802

File: extensions/test/algorithms/parse.cpp [hunk 2]
```diff
@@ -2,36 +2,38 @@
...
-	BOOST_CHECK_CLOSE(((double)bg::get<1>(geo_point_result)), expected_result_coord_2, 1.0e-6);
-
-	P p;
-	parse(p, coordinate_1, coordinate_2);
-	BOOST_CHECK_CLOSE(((double)bg::get<0>(p)), expected_result_coord_1, 1.0e-6);
-	BOOST_CHECK_CLOSE(((double)bg::get<1>(p)), expected_result_coord_2, 1.0e-6);
+	Geometry geometry;
+	geometry = parse<Geometry>(coordinate_1, coordinate_2);
+	BOOST_CHECK_CLOSE(((double)bg::get<0>(geometry)), expected_result_coord_1, 1.0e-4);
+	BOOST_CHECK_CLOSE(((double)bg::get<1>(geometry)), expected_result_coord_2, 1.0e-4);
```
Comment (mloskot) [inline extensions/test/algorithms/parse.cpp:32]: Such change from `1.0e-6` to `1.0e-4`  needs a word of comment, ideally, in commit message

File: extensions/test/algorithms/parse.cpp [hunk 3]
```diff
@@ -104,6 +106,26 @@ void test_all_complex_dms_strategy()
 
 	test_parse_with_point_and_parse_strategy<P, parser_t>(std::string("5.15R E"), std::string("5.15R N"),
 		expected_dms_result9, expected_dms_result10, parser_t());
+
+	// test with random characters in strategy(for generality check)
+	typedef bg::projections::detail::dms_parser<coord_t, true, 'A', 'B', 'C', 'D', 'm', 's', 'o', 'p'> parser_t1;
```
Comment (mloskot) [inline extensions/test/algorithms/parse.cpp:111]: Since we moved to C++14, we should prefer
```
using parser_t1 = bg::projections::detail::dms_parser<coord_t, true, 'A', 'B', 'C', 'D', 'm', 's', 'o', 'p'>;
```

which places the alias name to the left, it's easier to read.

File: extensions/test/algorithms/parse.cpp [hunk 4]
```diff
@@ -140,32 +162,49 @@ void test_all_with_simple_dms_strategy()
...
-		std::string("2S"), -1 * bg::math::d2r<double>(), -2 * bg::math::d2r<double>());
-	test_parse<bg::model::point<double, 2, bg::cs::spherical<bg::radian> > >(std::string("1W"),
-		std::string("2S"), -1 * bg::math::d2r<double>(), -2 * bg::math::d2r<double>());
-	test_parse<bg::model::point<double, 2, bg::cs::spherical_equatorial<bg::radian> > >(std::string("1W"),
-		std::string("2S"), -1 * bg::math::d2r<double>(), -2 * bg::math::d2r<double>());
-	test_parse<bg::model::point<double, 2, bg::cs::undefined> >(std::string("1W"),
-		std::string("2S"), -1 * bg::math::d2r<double>(), -2 * bg::math::d2r<double>());
-	test_parse<bg::model::point<double, 2, bg::cs::cartesian> >(std::string("1W"),
-		std::string("2S"), -1 * bg::math::d2r<double>(), -2 * bg::math::d2r<double>());
+	typedef typename bg::coordinate_type<Geometry>::type T;
```
Comment (mloskot) [inline extensions/test/algorithms/parse.cpp:171]: `using`

**Review by mloskot — CHANGES_REQUESTED** [review]

**Comment by mloskot** [comment]
@ayushgupta138 Please hit the "Resolve conversation" button for the comment threads that you have addressed.

File: extensions/test/algorithms/Jamfile (hunk 1 above)
Comment (ayushgupta138) [reply extensions/test/algorithms/Jamfile]: removed in new commit

File: extensions/test/algorithms/parse.cpp (hunk 2 above)
Comment (ayushgupta138) [reply extensions/test/algorithms/parse.cpp:32]: mentioned now

**Comment by ayushgupta138** [comment]
Actually I am new to contributing in orgs based settings. It will take some time to get used to all the protocols involved.

File: extensions/test/algorithms/parse.cpp (hunk 2 above)
Comment (mloskot) [reply extensions/test/algorithms/parse.cpp:32]: Yes, the commit https://github.com/boostorg/geometry/pull/806/commits/138d3d53e66a65dfd12be0af7815bc9493ed49fc  does mention it:

```
and reduce the error range from 1e-6 to 1e-4
```

but this is just *repeating **what** the commit does change* (i.e. it is equivalent to what everyone can see in the diff :-)), but it does not explain **why** the range changed. The **why** is what the commit messages are for. The **what** everyone can see by looking at the diff of a commit.

So, I'm hitting "Unresolve conversation" as this still needs clarification.

Comment (ayushgupta138) [reply extensions/test/algorithms/parse.cpp:32]: The range changed because I have also included test cases where the final value is very close to zero, so even though the difference between the actual value and expected value is of order 1e-9 , the percentage error exceed the previous error range of 1e-6 and had to be reduced to 1e-4. This was the case when 10 precision points were specified in the expected value. Is this explanation satisfying enough? If so I will add the explanation in the commit message after your confirmation.

Comment (mloskot) [reply extensions/test/algorithms/parse.cpp:32]: Yes, it is indeed.

**Review by mloskot — APPROVED** [review]

**Comment by mloskot** [comment]
@ayushgupta138 Thank you
